File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head == nullptr)
            return nullptr;
        
        ListNode* dummy = new ListNode(-1, head);
        
        ListNode* last = head;
        for(int i = 1; i < k and last != nullptr; i++) {
            last = last->next;
        }
        if(last == nullptr)
            return head;

        ListNode* nextPart = last->next;
        last->next = nullptr;

        reverseList(head);

        dummy->next = last;
        head->next = reverseKGroup(nextPart, k);

        return dummy->next;
        
    }

private:
    ListNode* reverseList(ListNode* head) {
        if(head == nullptr)
            return nullptr;
        ListNode* newHead = head;
        if(head->next != nullptr){
            newHead = reverseList(head->next);
            head->next->next = head;
        }
        head->next = nullptr;
        return newHead;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1, None)
        curr = dummy

        carry = 0
        while l1 or l2:
            s = 0 if not l1 else l1.val
            s = s + (0 if not l2 else l2.val)
            s = s + carry
            carry = s // 10
            s = s % 10
            curr.next = ListNode(s, None)
            curr = curr.next
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        
        if carry:
            curr.next = ListNode(carry, None)
        
        return dummy.next